### ISSUE TITLE
Provide k-means model trained from HUBERT_BASE, to further pre-train HUBERT_LARGE

### ISSUE THREAD

**Issue by arlofaria** [pytorch/audio#2883]
### 🚀 The feature

Would it be possible to share the k-means model that was used to produce targets for third-iteration pre-training of `HUBERT_LARGE` and `HUBERT_XLARGE` models?

This assumes that such a k-means model was trained on the packaged `HUBERT_BASE` model, and not some other variant of a nearly identical model of that same size, since the specific ordering of the centroids is critical.  If not, would it be possible to also share the corresponding `BASE`-sized model for the k-means model?

### Motivation, pitch

I'd be curious to experiment with _further pre-training_ of the `HUBERT_LARGE` model, i.e. generating features from the 9th layer of the provided `HUBERT_BASE` model applied to my own dataset, and then clustering with a pre-trained k-means model that is hopefully compatible between the 2nd and 3rd iterations.

The hope is that further pre-training on a large amount of unlabeled in-domain data might provide gains that are complementary to downstream fine-tuning on a small amount of labeled in-domain data.

### Alternatives

The alternative would be for me to produce my own k-means model for a `BASE`-sized HuBERT model, which is easy enough ... but then I wouldn't be able to do _further pre-training_ -- I'd have to start the 3rd iteration from scratch.  To produce the desired result, I'd then need to replicate FAIR's significant efforts to pre-train `HUBERT_LARGE` on 60K hours of LibriLight.

**Comment by nateanl** [comment]
Hi @arlofaria, thanks for sharing the issue. Let me confirm with authors of HuBERT to see whether the k-means model is still available.

**Comment by nateanl** [comment]
The desired k-means model is shared in Fairseq (https://github.com/facebookresearch/fairseq/tree/main/examples/hubert#pre-trained-and-fine-tuned-asr-models). You can download the "[L9 km500](https://dl.fbaipublicfiles.com/hubert/hubert_base_ls960_L9_km500.bin)" model which is trained on 9th layer of HuBERT BASE model in 2nd iteration.

**Comment by arlofaria** [comment]
Fantastic, thank you for the pointer! ... and sorry I didn't notice that it was published there already 🤦

**Comment by mthrok** [comment]
@nateanl is the model structure compatible with the one in torchaudio example dir?

**Comment by arlofaria** [comment]
From what I can tell, it appears to be compatible, as I can call `joblib.load` on it and get the following object:
```
MiniBatchKMeans(batch_size=10000, compute_labels=False, max_no_improvement=100,
                n_clusters=500, n_init=20, reassignment_ratio=0.5, verbose=1)
```

The only difference relative to the current HuBERT recipes in TorchAudio is that the `reassignment_ratio` defaults to 0.0.  That shouldn't matter for the purposes of k-means application, though I'd be curious if the difference in training was intentional.

However, there's also this warning:
```
UserWarning: Trying to unpickle estimator MiniBatchKMeans from version 0.24.0 when using version 1.1.3. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
```
... I'm not sure how much scikit-learn may have changed in the past 18 months, but seems unlikely that their k-means model would have changed in a meaningful way, right?

**Comment by nateanl** [comment]
the value of `reassignment_ratio ` is from https://github.com/facebookresearch/fairseq/blob/fcca32258c8e8bcc9f9890bf4714fa2f96b6b3e1/examples/hubert/simple_kmeans/learn_kmeans.py#L142, where it is 0.0 by default. Did you see other places that use 0.5 for training?

**Comment by nateanl** [comment]
I think the scikit-learn version should be fine. as only the numpy arrays of the cluster centers are used from the k-means model. You can use the api in https://github.com/pytorch/audio/blob/55e9978a68067f7f3b70ece667e25a12ed032a06/examples/hubert/utils/kmeans.py#L138-L151

**Comment by arlofaria** [comment]
> the value of `reassignment_ratio ` is from https://github.com/facebookresearch/fairseq/blob/fcca32258c8e8bcc9f9890bf4714fa2f96b6b3e1/examples/hubert/simple_kmeans/learn_kmeans.py#L142, where it is 0.0 by default. Did you see other places that use 0.5 for training?

Yes, apparently the default is 0.5 when used in the newer GLSM model for textless NLP:
https://github.com/facebookresearch/fairseq/blob/f131336fc303992cf309be3953bf523e1654fa1f/examples/textless_nlp/gslm/speech2unit/clustering/cluster_kmeans.py#L94

**Comment by jojonki** [comment]
I'd like to share my experience with `reassignment_ratio` in GSLM kmeans models. 

I downloaded the kmeans model from GSLM (https://medium.com/@aaltanim/constructing-a-clustering-model-employing-the-k-means-algorithm-546411fd9733) and confirmed that `reassignment_ratio` is set to `0.5`. I also trained LibriSpeech kmeans model using this setting, and the resulting model had a severe data imbalance in certain clusters.

As nateanl showed us, setting `reassignment_ratio` to `0.0` solves this problem. `reassignment_ratio` should generally be set to a very small value. `0.5` is unreasonably large and can lead to imbalanced clusters.

I suspect that the publicly available GSLM kmeans models were accidentally saved with `reassignment_ratio=0.5` after training.
